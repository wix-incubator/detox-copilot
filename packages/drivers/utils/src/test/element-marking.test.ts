import {
  TestContext,
  setupTestEnvironment,
  teardownTestEnvironment,
} from "./setup";

describe("Element Marking Functionality", () => {
  let testContext: TestContext;

  beforeAll(async () => {
    testContext = await setupTestEnvironment();
  }, 30000);

  afterAll(async () => {
    await teardownTestEnvironment(testContext);
  });

  beforeEach(async () => {
    // Reset state before each test
    await testContext.page.evaluate(() => {
      window.driverUtils.cleanupStyleChanges();
    });
  });

  it("should mark all supported elements with correct categories", async () => {
    const categories = await testContext.page.evaluate(() => {
      window.driverUtils.markImportantElements();
      const elements = document.querySelectorAll("[aria-pilot-category]");
      return Array.from(elements).map((el) => ({
        tag: el.tagName.toLowerCase(),
        category: el.getAttribute("aria-pilot-category"),
        index: el.getAttribute("aria-pilot-index"),
      }));
    });

    expect(categories).toMatchSnapshot("element-categories");
  });

  it("should handle hidden elements correctly", async () => {
    const hiddenElementsResult = await testContext.page.evaluate(() => {
      // Add some hidden elements
      const div = document.createElement("div");
      div.innerHTML = `
        <button style="display: none">Hidden Button</button>
        <a href="#" style="visibility: hidden">Hidden Link</a>
        <input type="text" hidden>
      `;
      document.body.appendChild(div);

      // Mark elements without including hidden
      window.driverUtils.markImportantElements();

      // Check if hidden elements were marked
      const hiddenElements = Array.from(div.children);
      return hiddenElements.map((el) => ({
        tag: el.tagName.toLowerCase(),
        hasCategory: el.hasAttribute("aria-pilot-category"),
      }));
    });

    expect(hiddenElementsResult).toMatchSnapshot("hidden-elements-handling");
  });

  it("should mark hidden elements when includeHidden option is true", async () => {
    const hiddenElementsResult = await testContext.page.evaluate(() => {
      window.driverUtils.markImportantElements({ includeHidden: true });
      const elements = document.querySelectorAll("[aria-pilot-category]");
      return Array.from(elements).map((el) => ({
        tag: el.tagName.toLowerCase(),
        category: el.getAttribute("aria-pilot-category"),
        isHidden:
          window.getComputedStyle(el).display === "none" ||
          window.getComputedStyle(el).visibility === "hidden" ||
          el.hasAttribute("hidden"),
      }));
    });

    expect(hiddenElementsResult).toMatchSnapshot("hidden-elements-included");
  });
});
