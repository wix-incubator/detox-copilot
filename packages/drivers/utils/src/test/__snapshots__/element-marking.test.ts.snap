// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element Marking Functionality should handle hidden elements correctly: hidden-elements-handling 1`] = `
[
  {
    "hasCategory": false,
    "tag": "button",
  },
  {
    "hasCategory": false,
    "tag": "a",
  },
  {
    "hasCategory": false,
    "tag": "input",
  },
]
`;

exports[`Element Marking Functionality should mark all supported elements with correct categories: element-categories 1`] = `
[
  {
    "category": "semantic",
    "index": "0",
    "tag": "header",
  },
  {
    "category": "header",
    "index": "0",
    "tag": "h1",
  },
  {
    "category": "semantic",
    "index": "1",
    "tag": "nav",
  },
  {
    "category": "list",
    "index": "0",
    "tag": "ul",
  },
  {
    "category": "link",
    "index": "0",
    "tag": "a",
  },
  {
    "category": "link",
    "index": "1",
    "tag": "a",
  },
  {
    "category": "link",
    "index": "2",
    "tag": "a",
  },
  {
    "category": "semantic",
    "index": "2",
    "tag": "main",
  },
  {
    "category": "semantic",
    "index": "3",
    "tag": "section",
  },
  {
    "category": "header",
    "index": "1",
    "tag": "h2",
  },
  {
    "category": "button",
    "index": "0",
    "tag": "button",
  },
  {
    "category": "button",
    "index": "1",
    "tag": "button",
  },
  {
    "category": "button",
    "index": "2",
    "tag": "button",
  },
  {
    "category": "input",
    "index": "0",
    "tag": "input",
  },
  {
    "category": "input",
    "index": "1",
    "tag": "input",
  },
  {
    "category": "input",
    "index": "2",
    "tag": "input",
  },
  {
    "category": "input",
    "index": "3",
    "tag": "input",
  },
  {
    "category": "input",
    "index": "4",
    "tag": "select",
  },
  {
    "category": "semantic",
    "index": "4",
    "tag": "section",
  },
  {
    "category": "header",
    "index": "2",
    "tag": "h2",
  },
  {
    "category": "list",
    "index": "1",
    "tag": "ul",
  },
  {
    "category": "list",
    "index": "2",
    "tag": "ol",
  },
  {
    "category": "semantic",
    "index": "5",
    "tag": "section",
  },
  {
    "category": "header",
    "index": "3",
    "tag": "h2",
  },
  {
    "category": "table",
    "index": "0",
    "tag": "table",
  },
  {
    "category": "semantic",
    "index": "6",
    "tag": "section",
  },
  {
    "category": "header",
    "index": "4",
    "tag": "h2",
  },
  {
    "category": "button",
    "index": "3",
    "tag": "button",
  },
  {
    "category": "link",
    "index": "3",
    "tag": "a",
  },
  {
    "category": "semantic",
    "index": "7",
    "tag": "footer",
  },
]
`;

exports[`Element Marking Functionality should mark hidden elements when includeHidden option is true: hidden-elements-included 1`] = `
[
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "header",
  },
  {
    "category": "header",
    "isHidden": false,
    "tag": "h1",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "nav",
  },
  {
    "category": "list",
    "isHidden": false,
    "tag": "ul",
  },
  {
    "category": "link",
    "isHidden": false,
    "tag": "a",
  },
  {
    "category": "link",
    "isHidden": false,
    "tag": "a",
  },
  {
    "category": "link",
    "isHidden": false,
    "tag": "a",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "main",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "section",
  },
  {
    "category": "header",
    "isHidden": false,
    "tag": "h2",
  },
  {
    "category": "button",
    "isHidden": false,
    "tag": "button",
  },
  {
    "category": "button",
    "isHidden": false,
    "tag": "button",
  },
  {
    "category": "button",
    "isHidden": false,
    "tag": "button",
  },
  {
    "category": "input",
    "isHidden": false,
    "tag": "input",
  },
  {
    "category": "input",
    "isHidden": false,
    "tag": "input",
  },
  {
    "category": "input",
    "isHidden": false,
    "tag": "input",
  },
  {
    "category": "input",
    "isHidden": false,
    "tag": "input",
  },
  {
    "category": "input",
    "isHidden": false,
    "tag": "select",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "section",
  },
  {
    "category": "header",
    "isHidden": false,
    "tag": "h2",
  },
  {
    "category": "list",
    "isHidden": false,
    "tag": "ul",
  },
  {
    "category": "list",
    "isHidden": false,
    "tag": "ol",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "section",
  },
  {
    "category": "header",
    "isHidden": false,
    "tag": "h2",
  },
  {
    "category": "table",
    "isHidden": false,
    "tag": "table",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "section",
  },
  {
    "category": "header",
    "isHidden": false,
    "tag": "h2",
  },
  {
    "category": "button",
    "isHidden": false,
    "tag": "button",
  },
  {
    "category": "link",
    "isHidden": false,
    "tag": "a",
  },
  {
    "category": "button",
    "isHidden": true,
    "tag": "button",
  },
  {
    "category": "input",
    "isHidden": true,
    "tag": "input",
  },
  {
    "category": "semantic",
    "isHidden": false,
    "tag": "footer",
  },
  {
    "category": "button",
    "isHidden": true,
    "tag": "button",
  },
  {
    "category": "link",
    "isHidden": true,
    "tag": "a",
  },
  {
    "category": "input",
    "isHidden": true,
    "tag": "input",
  },
]
`;
